network/server/tchannelthrift/pool.go

```diff
@@ -0,0 +1,161 @@
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package tchannelthrift
+
+import (
+	"github.com/m3db/m3db/generated/thrift/rpc"
+	"github.com/m3db/m3db/pool"
+)
+
+var (
+	emptyBlockMetadata  rpc.BlockMetadata
+	emptyBlocksMetadata rpc.BlocksMetadata
+)
+
+// BlockMetadataPool provides a pool for block metadata
+type BlockMetadataPool interface {
+	// Get returns a block metadata
+	Get() *rpc.BlockMetadata
+
+	// Put puts a block metadata back to pool
+	Put(m *rpc.BlockMetadata)
+}
+
+// BlockMetadataSlicePool provides a pool for block metadata slices
+type BlockMetadataSlicePool interface {
+	// Get returns a block metadata slice
+	Get() []*rpc.BlockMetadata
+
+	// Put puts a block metadata slice back to pool
+	Put(m []*rpc.BlockMetadata)
+}
+
+// BlocksMetadataPool provides a pool for blocks metadata
+type BlocksMetadataPool interface {
+	// Get returns a blocks metadata
+	Get() *rpc.BlocksMetadata
+
+	// Put puts a blocks metadata back to pool
+	Put(m *rpc.BlocksMetadata)
+}
+
+// BlocksMetadataSlicePool provides a pool for blocks metadata slices
+type BlocksMetadataSlicePool interface {
+	// Get returns a blocks metadata slice
+	Get() []*rpc.BlocksMetadata
+
+	// Put puts a blocks metadata slice back to pool
+	Put(m []*rpc.BlocksMetadata)
+}
+
+type blockMetadataPool struct {
+	pool pool.ObjectPool
+}
+
+// NewBlockMetadataPool creates a new block metadata pool
+func NewBlockMetadataPool(opts pool.ObjectPoolOptions) BlockMetadataPool {
+	p := pool.NewObjectPool(opts)
+	p.Init(func() interface{} {
+		return rpc.NewBlockMetadata()
+	})
+	return &blockMetadataPool{pool: p}
+}
+
+func (p *blockMetadataPool) Get() *rpc.BlockMetadata {
+	return p.pool.Get().(*rpc.BlockMetadata)
+}
+
+func (p *blockMetadataPool) Put(metadata *rpc.BlockMetadata) {
+	if metadata != nil {
+		*metadata = emptyBlockMetadata
+		p.pool.Put(metadata)
+	}
+}
+
+type blockMetadataSlicePool struct {
+	pool     pool.ObjectPool
+	capacity int
+}
+
+// NewBlockMetadataSlicePool creates a new blockMetadataSlice pool
+func NewBlockMetadataSlicePool(opts pool.ObjectPoolOptions, capacity int) BlockMetadataSlicePool {
+	p := pool.NewObjectPool(opts)
+	p.Init(func() interface{} {
+		return make([]*rpc.BlockMetadata, 0, capacity)
+	})
+	return &blockMetadataSlicePool{pool: p, capacity: capacity}
+}
+
+func (p *blockMetadataSlicePool) Get() []*rpc.BlockMetadata {
+	return p.pool.Get().([]*rpc.BlockMetadata)
+}
+
+func (p *blockMetadataSlicePool) Put(res []*rpc.BlockMetadata) {
+	res = res[:0]
```

Comment:
Should we defend against putting extremely large slices back in the pool if they grew outside their capacity?